### ISSUE TITLE
Provide a data structure merging static and dynamic data

### ISSUE THREAD

**Issue by echebbi** [gemoc/ale-lang#140]
## Motivation

When using ALE's Java API to manipulate an AST we usually want to deal with static and dynamic attributes/methods uniformly. In other words, we don't care whether an attribute has been declared in Ecore or within an ALE source file, we just want to check its name/type/value. However, static and dynamic data are currently split in two: 

 - static data are represented by `EClass` instances,
 - dynamic data are represented by `BehavioredClass` instances.

As a result, and to make sure both are taken into account, we usually have to do things twice (examples: for [providing autocomplete](https://github.com/gemoc/ale-lang/blob/88956d2785c788f0da377bf3140b44384235f99e/plugins/org.eclipse.emf.ecoretools.ale.xtext.ui/src/org/eclipse/emf/ecoretools/ui/contentassist/AleProposalProvider.xtend#L103-L121) and for [type checking](https://github.com/gemoc/ale-lang/blob/88956d2785c788f0da377bf3140b44384235f99e/plugins/org.eclipse.emf.ecoretools.ale.core/src/org/eclipse/emf/ecoretools/ale/core/validation/impl/AstLookup.java#L153-L182)). This is also outlined by the fact that `IAleEnvironment` proposes both a `getMetamodels()` and a `getBehaviors()` method.

Besides the redundancy, this makes the code more complex, harder to read, to write and to maintain.

## Requirements

The solution must:
 - allow to access both static and dynamic data
 - provide access to static and dynamic data of the same kind (class, attributes, operations, etc.) in a uniformly, purely transparent way (think polymorphism)
 - allow to check whether a data is static or dynamic (e.g. `clazz.methods.get("foo()").isStatic()`): this is required at least to distinguish data in Sirius' *Behaviors* layer.

## Proposition

I see 3 options:

1. Turn every `BehavioredClass` into an `EClass`
2. Enhance `BehavioredClass` to also provide static data
3. Create a third representation that mixes both static and dynamic data.

I'm against ① because:
 - `EClass` are not objects but data bags, harming APIs (this is a personal take though),
 - we cannot change the implementation/contract of `EClass` & cie, which would likely make language evolutions hard,
 - as a result of the latter we would likely have to store some info as EAnnotation, which would be unbearable.

I can't decide between the other options because:
 - ② is by far the simpler to implement and to maintain
 - but ③ would allow for cleaner API and would make easier to evolve the language independently from the API, and vice versa.

Well, actually we're back to the same problem I face with every EMF-based project. My personal preference would be ③, but I believe ② is better for the project.

Modifications should anyway be considered with care to prevent important regressions.

## Related issue

Feels related to https://github.com/gemoc/ale-lang/issues/123#issuecomment-623161170 (provide an API to manipulate runtime data).

**Comment by dvojtise** [comment]
Distinction between static and dynamic is also required to correctly set up the recording of the runtime data trace (typical usecase of the timeline in GEMOC, this trace stores only dynamic information since they are the only one that can change during model execution)


I like solution ③, it makes me feel like a kind of abstract view on top of the technical implementation, where is shows the model resulting from the merge of all sources (static eclass and possibly several ALE aspect classes). Its API would primarily focus on the resulting features available for a given class. inspecting where does a specific feature comes from (ecore or a given ale aspect) is a secondary concerns.
Additionnaly, it can also abstract the idea that in ecore too we can "tag" a given attribute/class to be considered as "dynamic"

**Comment by echebbi** [comment]
> I like solution ③, it makes me feel like a kind of abstract view on top of the technical implementation

Absolutely. And a nice side effect is that we could introduce this new API without having to change the way `BehavioredClass` instances are built. My main concerns are that:
 - we would end up with _yet another_ set of `Class`, `Attribute` and `Method` classes,
 - the `(EClass + BehavioredClass) -> ComprehensiveClass` adapter may be cumbersome to write and maintain,
 - we would have to replace a significant part of the code manipulating `BehavioredClass` instances.

I believe that it would be for the better (resulting code may be significantly cleaner) but that's a lot of work, susceptible to introduce regressions, and I'm not sure we have the bandwidth to handle it at the moment.
